### PR TITLE
[CPDLP-3131] Add participant id changes table

### PR DESCRIPTION
**Pull request by leandroalemao**
### Context

Ticket: [CPDLP-3131](https://dfedigital.atlassian.net/browse/CPDLP-3131)

We will need to add the participant id changes table to surface the correct structure in the participants npq endpoint in the API.

### Changes proposed in this pull request

Add participant id changes table with correct fields/validations.

[CPDLP-3131]: https://dfedigital.atlassian.net/browse/CPDLP-3131?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/models/participant_id_change.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class ParticipantIdChange < ApplicationRecord
+  has_paper_trail
```
Comment (cwrw) [inline app/models/participant_id_change.rb:4]: could we postpone adding papertrail for now, we will do a sweep in the end to add it to an audited list of tables if that's ok?

File: spec/factories/applications.rb [hunk 2]
```diff
@@ -81,5 +81,13 @@
     trait :with_random_user do
       user { FactoryBot.build(:user, :with_random_name) }
     end
+
+    trait :with_participant_id_change do
+      after(:create) do |application|
+        user = application.user
+
+        FactoryBot.create(:participant_id_change, from_participant: user, user:)
```
Comment (cwrw) [inline spec/factories/applications.rb:89]: could we amend this to move from a previous other user to the current user? maybe something like this:
`FactoryBot.create(:participant_id_change, to_participant: user, user:)`

**Review by cwrw — APPROVED** [review]
looking good just some minor comments

**Comment by peteryates** [comment]
Looks good.

A couple of questions.

I'm slightly confused about how we use/surface this data, and does tracking the entire history of changes introduce querying complexity?

i.e. if a participant's changed more than once to get all their previous ids we'd need a recursive query:

```mermaid
flowchart TD
    123 --> 234
    234 --> 345
```

Like here, `345` used to be both `123` and `234`.

Would we be at a disadvantage if we just saved the `former_participant_ids` in an array?

```yaml
Participant:
  id: 345
  former_ids: [123, 234]
```

Do we need to know when the changes happened?

**Comment by cwrw** [comment]
> i.e. if a participant's changed more than once to get all their previous ids we'd need a recursive query

We link everything to the current user, so we just have to look at the current user and list the participant id changes 
```
object.participant_id_changes
```

> Would we be at a disadvantage if we just saved the former_participant_ids in an array?

We do surface the `changed_at` in serialisers so we can't lose that data unfortunately

**Comment by peteryates** [comment]
Ok - I missed that but that means we have 3 foreign keys on `participant_id_changes` pointing at the `users.id`, and two end with `participant_id`.

It wasn't clear to me what they meant and there's no other mention of 'participant' in the database.

**Comment by cwrw** [comment]
> Should it be user_id_changes?
Good question, I think we went with the participant in the name as it maps identically to what's in the API and avoids confusion for us.

**Pull request merged**